src/Processors/Executors/CompletedPipelineExecutor.cpp

```diff
@@ -72,9 +72,11 @@ void CompletedPipelineExecutor::execute()
         data->executor = std::make_shared<PipelineExecutor>(pipeline.processors, pipeline.process_list_element);
         data->executor->setReadProgressCallback(pipeline.getReadProgressCallback());
 
-        auto func = [&, thread_group = CurrentThread::getGroup()]()
+        /// Avoid passing this to labmda, copy ptr to data instead.
+        /// Destructor of unique_ptr copy raw ptr into local variable first, only then calls object destructor.
+        auto func = [data_ptr = data.get(), num_threads = pipeline.getNumThreads(), thread_group = CurrentThread::getGroup()]()
         {
-            threadFunction(*data, thread_group, pipeline.getNumThreads());
+            threadFunction(*data_ptr, thread_group, num_threads);
```

Comment:
This seems even more questionable with accessing raw pointer of `unique_ptr`. 
You can still access invalid memory if `Data` is deleted, right?

Reply:
> You can still access invalid memory if Data is deleted, right?

No. Because we are waiting for the thread to finish in `Data` destructor.
Memory is valid until `~Data` is finished, thread will be joined before ~Data is finished.
The race was on the field inside `unique_ptr` itself, now it should not happen.

> you want to wait for the data.thread to join in `CompletedPipelineExecutor` destructor and not in the `Data` destructor.

It is kind of the same right now. Just, while joining thread in `~Data` I don't have to think about exception safety.

Reply:
`~Data()` is responsible for joining it's thread. The invariant is quite simple : the thread is joined after `~Data`.
Thread itself can change only the fields inside `Data`. This invariant was broken (cause we used captured this), now it is fixed.

Reply:
you're correct, it makes more sense like this